### ISSUE TITLE
Intermediate supervision

### ISSUE THREAD

**Issue by visonpon** [bearpaw/pytorch-pose#13]
thanks for the pytorch version,  it seems that adding the intermediate loss to the single hourglass haven't repreduce?
btw, is all the intermediate losses added together at last and then do bp?

**Comment by bearpaw** [comment]
Hi @visonpon , what do you mean by adding the intermediate loss the single hourglass model? The model in this code is exactly the same as the original ECCV16 paper, where intermediate losses are added to the end of each stack of hourglass. For example, 8-stack hourglass model has 8 losses.

**Comment by visonpon** [comment]
@bearpaw I know that every hourglass have a loss, i mean is every hourglass updated independently according to the their loss or all this losses added together and then do BP to update the whole stacked netwroks?thanks~

**Comment by bearpaw** [comment]
Each hourglass should be updated according to its own loss, as in https://github.com/bearpaw/pytorch-pose/blob/master/example/mpii.py#L146-L148

**Comment by visonpon** [comment]
@bearpaw So every hourglass trained independently and then stacked to output the final result? so this is not an end-to-end training?

**Comment by bearpaw** [comment]
It's end-to-end training. The gradient of the lower stack comes from two sources: the gradients from its own intermediate loss, and the gradients back-propagated from the higher stacks.

**Comment by bearpaw** [comment]
Closed for no more discussion.
